src/Elastic.Apm.StackExchange.Redis/ElasticApmProfiler.cs

```diff
@@ -23,13 +23,13 @@ internal class ElasticApmProfiler
 		private readonly ConcurrentDictionary<string, ProfilingSession> _executionSegmentSessions =
 			new ConcurrentDictionary<string, ProfilingSession>();
 
-		private readonly IApmLogger _logger;
-		private readonly IApmAgent _agent;
+		private readonly Lazy<IApmLogger> _logger;
+		private readonly Lazy<IApmAgent> _agent;
 
-		public ElasticApmProfiler(IApmAgent agent)
+		public ElasticApmProfiler(Lazy<IApmAgent> agent)

```

Comment:
Instead of passing a lazy here, maybe have it take a factory function `Func<IApmAgent>` which is what `Lazy` takes. Seems weird to me to have to do `new ElasticApmProfiler(new Lazy<IApmAgent>(() => Agent.Instance))` instead of just `new ElasticApmProfiler(() => Agent.Instance)` which you can then add to your internal Lazy.

Reply:
a `Func<IApmAgent>` is a bit cleaner 👍